adapters/beintoo/beintoo.go

```diff
@@ -0,0 +1,254 @@
+		addHeaderIfNonEmpty(headers, "Referer", request.Site.Page)
+	}
+
+	url := buildEndpoint(a.endpoint, a.testing, request.TMax)
+
+	return []*adapters.RequestData{{
+		Method:  "POST",
+		Uri:     url,
+		Body:    data,
+		Headers: headers,
+	}}, errs
+}
+
+func unpackImpExt(imp *openrtb.Imp) (*openrtb_ext.ExtImpBeintoo, error) {
+	var bidderExt adapters.ExtImpBidder
+	if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
+		return nil, &errortypes.BadInput{
+			Message: err.Error(),
+		}
+	}
+
+	var BeintooExt openrtb_ext.ExtImpBeintoo
+	if err := json.Unmarshal(bidderExt.Bidder, &BeintooExt); err != nil {
+		return nil, &errortypes.BadInput{
+			Message: fmt.Sprintf("ignoring imp id=%s, invalid ImpExt", imp.ID),
+		}
+	}
+
+	tagIDValidation, err := strconv.ParseInt(BeintooExt.TagID, 10, 64)
+	if err != nil || tagIDValidation == 0 {
+		return nil, &errortypes.BadInput{
+			Message: fmt.Sprintf("ignoring imp id=%s, invalid tagid must be a String of numbers", imp.ID),
+		}
+	}
+
+	return &BeintooExt, nil
+}
+
+func buildImpBanner(imp *openrtb.Imp) error {
+	imp.Ext = nil
+
+	if imp.Banner == nil {
+		return &errortypes.BadInput{
+			Message: fmt.Sprintf("Request needs to include a Banner object"),
+		}
+	}
+
+	bannerCopy := *imp.Banner
+	banner := &bannerCopy
+
+	if banner.W == nil && banner.H == nil {
+		if len(banner.Format) == 0 {
+			return &errortypes.BadInput{
+				Message: fmt.Sprintf("Need at least one size to build request"),
+			}
+		}
+		format := banner.Format[0]
+		banner.Format = banner.Format[1:]
+		banner.W = &format.W
+		banner.H = &format.H
+		imp.Banner = banner
+	}
+
+	return nil
+}
+
+// Add Beintoo required properties to Imp object
+func addImpProps(imp *openrtb.Imp, secure *int8, BeintooExt *openrtb_ext.ExtImpBeintoo) {
+	imp.TagID = BeintooExt.TagID
+	imp.Secure = secure
+
+	if BeintooExt.BidFloor != "" {
+		bidFloor, err := strconv.ParseFloat(BeintooExt.BidFloor, 64)
+		if err != nil {
+			bidFloor = 0
+		}
+
+		if bidFloor > 0 {
+			imp.BidFloor = bidFloor
+			imp.BidFloorCur = "USD"
+		}
+	}
+
+	return
+}
+
+// Adding header fields to request header
+func addHeaderIfNonEmpty(headers http.Header, headerName string, headerValue string) {
+	if len(headerValue) > 0 {
+		headers.Add(headerName, headerValue)
+	}
+}
+
+// Handle request errors and formatting to be sent to Beintoo
+func preprocess(request *openrtb.BidRequest) []error {
+	errors := make([]error, 0, len(request.Imp))
+	resImps := make([]openrtb.Imp, 0, len(request.Imp))
+	secure := int8(0)
+
+	if request.Site != nil && request.Site.Page != "" {
+		pageURL, err := url.Parse(request.Site.Page)
+		if err == nil && pageURL.Scheme == "https" {
+			secure = int8(1)
+		}
+	}
+
+	for _, imp := range request.Imp {
+		BeintooExt, err := unpackImpExt(&imp)
+		if err != nil {
+			errors = append(errors, err)
+			continue
+		}
+
+		addImpProps(&imp, &secure, BeintooExt)
+
+		if err := buildImpBanner(&imp); err != nil {
+			errors = append(errors, err)
+			continue
+		}
```

Comment:
Function `preprocess` calls `continue` everytime it finds an error looking for the good entries in the `request.Imp` array. But the `if` statement in line 57 is zero tolerant for any errors making the `continue` statements unnecessary:
```
 48 func (a *BeintooAdapter) MakeRequests(request *openrtb.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
 49     var errs []error
 50
 51     if len(request.Imp) == 0 {
 52         return nil, []error{&errortypes.BadInput{
 53             Message: fmt.Sprintf("No Imps in Bid Request"),
 54         }}
 55     }
 56
 57     if errs := preprocess(request); errs != nil && len(errs) > 0 { //<-- exits even if only one error was found
 58         return nil, append(errs, &errortypes.BadInput{
 59             Message: fmt.Sprintf("Error in preprocess of Imp, err: %s", errs),
 60         })
 61     }
 62
adapters/beintoo/beintoo.go
```
Under this logic, we might as well:
```
177 - func preprocess(request *openrtb.BidRequest) []error {
    + func preprocess(request *openrtb.BidRequest) error {
178 -     errors := make([]error, 0, len(request.Imp))
179       resImps := make([]openrtb.Imp, 0, len(request.Imp))
180       secure := int8(0)
181  
182   *--  6 lines: if request.Site != nil && request.Site.Page != "" {-----
188  
189       for _, imp := range request.Imp {
190           BeintooExt, err := unpackImpExt(&imp)
191           if err != nil {
192 -             errors = append(errors, err)
193 -             continue
    +             return err
194           }
195  
196           addImpProps(&imp, &secure, BeintooExt)
197  
198           if err := buildImpBanner(&imp); err != nil {
199 -             errors = append(errors, err)
200 -             continue
    +             return err
201           }
202           resImps = append(resImps, imp)
203       }
204  
205       request.Imp = resImps
206  
207 -     return errors
    +     return nil
208   }
adapters/beintoo/beintoo.go
```
Is this what we intended?